### PR TITLE
EI25Q3-005 - Grouping suggestions on the inspector (LEAN-4102)

### PR DESCRIPTION
**Pull request by asouqi**
This PR will group adjacent inline node suggestions in one change

### CODE REVIEW

File: src/changes/findChanges.ts [hunk 1]
```diff
@@ -47,21 +47,6 @@ export function findChanges(state: EditorState) {
...
-        current.change.id === id &&
-        current.node.isText &&
-        node.isText &&
-        !equalMarks(node, current.node)
-      ) {
-        current.change.to = pos + node.nodeSize
-        // Important to update the node as the text changes might contain multiple parts where some marks equal each other
-        current.node = node
-        continue
-      }
```
Comment (asouqi) [inline src/changes/findChanges.ts:64]: I removed this as that will keep the same change Id for text change. and we are going to do the join from changeTree for that case

**Review by mbartenev-atypon — CHANGES_REQUESTED** [review]

File: src/types/change.ts [hunk 2]
```diff
@@ -95,13 +95,18 @@ export type ReferenceChange = Change & {
...
+  nodes: TrackedChange[]
+}
 export type TrackedChange =
   | TextChange
   | NodeChange
   | NodeAttrChange
   | WrapChange
   | ReferenceChange
   | MarkChange
+  | InlineAdjacentChanges
```
[the remark this answers is not in the document]
Comment (asouqi) [reply src/types/change.ts:109]: what about adding a new method in `ChangeSet` to do the custom grouping we want and later composite changes? as we don't want to add anything in the document for that case or similar thing. also with that there will be no change in the type

[the remark this answers is not in the document]
Comment (mbartenev-atypon) [reply src/types/change.ts:109]: Sounds good, lets try that.

**Review by mbartenev-atypon — CHANGES_REQUESTED** [review]

File: src/types/change.ts [hunk 3]
```diff
@@ -95,13 +95,18 @@ export type ReferenceChange = Change & {
...
+  nodes: TrackedChange[]
+}
 export type TrackedChange =
   | TextChange
   | NodeChange
   | NodeAttrChange
   | WrapChange
   | ReferenceChange
   | MarkChange
+export type GroupedChange = TrackedChange | InlineAdjacentChanges
```
[the remark this answers is not in the document]
Comment (asouqi) [reply src/types/change.ts:109]: > What are the cases when GroupedChange can be just a TrackedChange?

if we have signal changes in the document.

I'm not sure how can we make them separate, as we can have a case of document contain a single change and group of changes how can we represent them in ChangeSet?

File: src/types/change.ts [hunk 4]
```diff
@@ -95,13 +95,18 @@ export type ReferenceChange = Change & {
 export type MarkChange = Change & {
   type: 'mark-change'
 }
+export type InlineAdjacentChanges = Change & {
```
[the remark this answers is not in the document]
Comment (asouqi) [reply src/types/change.ts:98]: updated to omit `dataTracked`, as we need `id, from, to`

File: src/types/change.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (asouqi) [reply src/types/change.ts]: updated type to:
```
export type SingleChange = Omit<Change, 'dataTracked'> & {
  type: 'single-change'
  node: TrackedChange
}
export type GroupedChange = SingleChange | InlineAdjacentChanges
```
@mbartenev-atypon what do you think about that?

[the remark this answers is not in the document]
Comment (mbartenev-atypon) [reply src/types/change.ts]: basically there should be some sort of FinalChange or SemanticChange that would represent a processed change or changes with which inspector or anything else can work without knowing such details as whether it is a group of changes or is it a single change in our node tree etc.

[the remark this answers is not in the document]
Comment (asouqi) [reply src/types/change.ts]: updated to be same type

File: src/ChangeSet.ts [hunk 5]
```diff
@@ -62,17 +62,29 @@ export class ChangeSet {
    * and end position. This is useful for showing the changes as groups in the UI.
    */
   get changeTree() {
-    const rootNodes: TrackedChange[] = []
+    const rootNodes: TrackedChange[][] = []
```
Comment (mbartenev-atypon) [inline src/ChangeSet.ts:65]: same comment as in article editor about covering the array inside the type

File: src/ChangeSet.ts [hunk 6]
```diff
@@ -84,7 +96,7 @@ export class ChangeSet {
         const result = this.matchAndAddToRootChange(rootNodes, c)
         if (result) {
           const { index, root } = result
-          rootNodes[index] = root
+          rootNodes[index][0] = root
```
Comment (mbartenev-atypon) [inline src/ChangeSet.ts:99]: please make a note somewhere that item [0] is considered a main change

Comment (asouqi) [reply src/ChangeSet.ts:102]: updated

File: test/apply-changes/apply-changes.test.ts [hunk 7]
```diff
@@ -106,9 +106,9 @@ describe('apply-changes.test', () => {
     }
 
     expect(tester.toJSON()).toEqual(insertAccept[1])
-    expect(uuidv4Mock.mock.calls.length).toBe(26)
+    expect(uuidv4Mock.mock.calls.length).toBe(27)
     expect(tester.trackState()?.changeSet.hasInconsistentData).toEqual(false)
-    expect(log.warn).toHaveBeenCalledTimes(0)
+    expect(log.warn).toHaveBeenCalledTimes(1)
```
[the remark this answers is not in the document]
Comment (asouqi) [reply test/apply-changes/apply-changes.test.ts]: fix that, reverted

File: src/ChangeSet.ts (hunk 5 above)
Comment (asouqi) [reply src/ChangeSet.ts]: updated types to:
```
export type RootChanges = TrackedChange[][]
export type RootChange = TrackedChange[]
```

**Pull request merged**